runtime/src/fees/payment.rs

```diff
@@ -13,24 +13,119 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{fees::Fee, AccountId};
+use crate::registry::store;
+use crate::{
+    fees::{bid::Bid, Fee},
+    AccountId, DispatchError, RegistryCall,
+};
+use radicle_registry_core::*;
+
+use frame_support::storage::StorageMap as _;
 use frame_support::traits::{Currency, ExistenceRequirement};
 
-/// Pay a given fee by withdrawing it from the `payee` account
-/// and transfering it, with a small burn, to the block author.
-pub fn pay_fee(fee: impl Fee, payee: &AccountId) {
-    // 1. Withdraw from payee
-    let _negative_imbalance = <crate::Balances as Currency<_>>::withdraw(
+type NegativeImbalance = <crate::Balances as Currency<AccountId>>::NegativeImbalance;
+
+/// Pay Fees
+/// Given a tx author, their bid, and a RegistryCall they are submitting,
+/// charge the tx fees to the right account, which depends on the `registry_call`.
+pub fn new_pay_fee(
+    author: AccountId,
+    bid: Bid,
+    registry_call: RegistryCall,
+) -> Result<(), DispatchError> {
+    let payee = decide_payee(author, registry_call);
+    withdraw_fee(bid, &payee)
+        .map(burn)
+        .and_then(pay_block_author)?;
```

Comment:
We’re mixing imperative and functional style here. I would prefer imperative style because it is easier to read and more idiomatic
```suggestion
    let withdrawn_fee = withdraw_fee(bid, &payee)?;
    let reward = burn(with_drawn_fee);
    pay_block_author(reward);
```

Reply:
https://media.giphy.com/media/d7rvF20PqNuGKSQGhf/giphy.gif

No, but I like your suggestion better too 👍